wallet-new/docs/updates.md

```diff
@@ -0,0 +1,656 @@
+
+The way updates will transition from the *next* to the *current* can be the following:
+
+- When a new update arrives, store it as the _next_ update. If there is already an
+  update designed as next, replace it.
+
+- When `GET /api/v1/update` is called:
+   - If the _current_ update is in the `Skipped` or `Applied` state, and a _next_ update is available,
+     then replace _current_ with _next_. *In-transit* `POST/PUT` http-requests won't
+     cause state inconsistencies by the virtue of the fact the input
+     `WalletSoftwareUpdate` won't match anymore.
+
+   - If the _current_ update is in the `Proposed` state, and a _next_ update is available,
+     then replace _current_ with _next_. *In-transit* `POST/PUT` http-requests won't
+     cause state inconsistencies by the virtue of the fact the input
+     `WalletSoftwareUpdate` won't match anymore.
+
+   - If the _current_ update is in the `Downloaded` state, and a _next_ update is available,
+     chances are we want to return the current (downloaded but not applied update) update so
+     that Daedalus can warn the user it has a downloaded (but not applied) update.
+
+- When `POST /api/v1/update` is called:
+    - If the input `WalletSoftwareUpdate` doesn't match the _current_ update, ignore the request.
+
+    - If the input `WalletSoftwareUpdate` does match the _current_ update, carry on as per
+      user story (4) and (4.1), then set the _current_ status to `Applied`.
+
+        - If a _next_ update is available, replace _next_ with _current_.
+
+## V1 Swagger spec<a id="sec-5-6" name="sec-5-6"></a>
+
+A mockup for the new update API might look like this (some HTTP errors and uninteresting
+types omitted for brevity):
+
+    swagger: '2.0'
+    info:
+      version: 1.0.2
+      title: V0 Update API
+      description: This is a subset of the API for Cardano SL wallet.
+    host: 'localhost:8090'
+    # Note how we still keep the singular name for the resource, as
+    # described here: http://www.restapitutorial.com/media/RESTful_Best_Practices-v1_1.pdf
+    # in the section "Pluralization (singleton resources)".
+    paths:
+      /api/v1/update:
+        get:
+          description: >
+            Returns the latest available update.
+            It's Daedalus' responsibility to check whether the returned update is
+            mandatory (at startup) and stop the user by using the Wallet. Note: this
+            might not be enough in case of clients different from Daedalus, as this is
+            an invariant which is not imposed anywhere if not on paper.
+          produces:
+            - application/json;charset=utf-8
+          responses:
+            '200':
+              schema:
+                $ref: '#/definitions/WalletSoftwareUpdate'
+              description: 'The latest available update.'
+        post:
+          description: >
+            Applies this update. It takes as input in the body
+            a JSON representing the `WalletSoftwareUpdate` to perform.
```

Comment:
This seems like a bit of an overkill. It's sufficient to provide version number or update hash -- either clearly identify an unique update proposal. I believe that even if we go that way, internally we'll compare input update to current one by version number or hash anyway, not as a whole (e. g. to avoid differences in `UpdateState`).

Reply:
I see what you mean, and I agree that it feels a bit heavy, but at the same time it's easier for the frontend, because they can simply re-post what they have got from `GET /api/v1/updates`. One middle ground could be to return the computed hash (let's call it `UpdateId`) as part of a `WalletSoftwareUpdate`, so that the frontend chaps can then extract it from the JSON and pass it to this post endpoint, like `POST /api/updates/{updateId}`.

Do you have any strong preference on this one?

Reply:
We already have such hash in `WalletSoftwareUpdate`, it's `wsuSha256Checksum`. I don't think we should invent something new.

I see no difference though between using hash or software version, because software version is also a part of `WalletSoftwareUpdate`. Though hash seems to be a little more precise (there shouln't exist two different updates with the same version, but who knows). 

I'm for using either software version or checksum hash for these endpoints instead of the whole datatype. I don't have any strong preference between these two options.

Reply:
> We already have such hash in WalletSoftwareUpdate, it's wsuSha256Checksum. I don't think we should invent something new.

@flyingleafe agreed, with the caveat that I have introduced this type as part of the proposal, and there might be extra work (which might not happen for weeks) to include such checksum in an `UpdateProposal` from the core layer, whereas we could go straight to my proposal and maybe still defer such "data integrity" decisions for later. Lame reason or did I convince you? 😀